test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc

```diff
@@ -483,6 +486,32 @@ TEST(QuicPlatformTest, QuicTestOutput) {
                       QuicRecordTestOutput("quic_test_output.3", "output 3 content\n"));
 }
 
+TEST(QuicPlatformTest, ConstructMemSliceFromBuffer) {
+  std::string str(512, 'b');
+  // Fragment needs to out-live buffer.
+  Envoy::Buffer::BufferFragmentImpl fragment(
+      str.data(), str.length(),
+      [](const void*, size_t, const Envoy::Buffer::BufferFragmentImpl*) {});
+  Envoy::Buffer::OwnedImpl buffer;
+  EXPECT_DEBUG_DEATH(quic::QuicMemSlice slice0{quic::QuicMemSliceImpl(buffer, 0)}, "");
+  std::string str2(1024, 'a');
+  buffer.add(str2);
+  EXPECT_EQ(1u, buffer.getRawSlices(nullptr, 0));
+  buffer.addBufferFragment(fragment);
+
+  quic::QuicMemSlice slice1{quic::QuicMemSliceImpl(buffer, str2.length())};
+  EXPECT_EQ(str.length(), buffer.length());
+  EXPECT_EQ(str2, std::string(slice1.data(), slice1.length()));
+  slice1.Reset();
+  EXPECT_TRUE(slice1.empty());
+  EXPECT_EQ(nullptr, slice1.data());
+
+  quic::QuicMemSlice slice2{quic::QuicMemSliceImpl(buffer, str.length())};
+  EXPECT_EQ(0, buffer.length());
+  EXPECT_EQ(str.data(), slice2.data());
+  EXPECT_EQ(str, std::string(slice2.data(), slice2.length()));
```

Comment:
Also verify that after slice2 destructs, the "done" callback of "fragment" is executed?